README.md
# Sy Notes

Sy's notes for Linkedin restli.

## Getting started

- https://linkedin.github.io/rest.li/get_started/quick_start

### Remote Debugging

Refer to [this stackoverflow for more information about remote debugging](Source: https://stackoverflow.com/questions/37702073/gradle-remote-debugging-process)

- Note The `"suspend=y"` part will pause the execution for you to attach a debugger.
- Start the server as usual but add the following param, so the command will look like this:

```
./gradlew startExampleBasicServer -Dorg.gradle.jvmargs='-Xdebug -Xrunjdwp:transport=dt_socket,server=y,address=5005,suspend=y'
```

```
./gradlew startExampleBasicClient -Dorg.gradle.jvmargs='-Xdebug -Xrunjdwp:transport=dt_socket,server=y,address=5005,suspend=y'
```

or in the `~/.gradle/gradle.properties`

```
org.gradle.daemon=false
org.gradle.jvmargs=-Xdebug -Xrunjdwp:transport=dt_socket,server=y,address=5005,suspend=y
```

Then in your InteliJ, add this configuration:

Use `Remote JVM Debug` and add this into the command line argument
`-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005`
![image](https://user-images.githubusercontent.com/3792401/155806095-ce429e57-cd6b-4d91-b421-521be5301c83.png)

##### Sample VS Code Launch Config

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "java",
      "name": "Remote Debugger Attach",
      "request": "attach",
      "hostName": "localhost",
      "port": "5005"
    }
  ]
}
```

## Requirements

```bash
>>> java -version
openjdk version "1.8.0_322"
OpenJDK Runtime Environment (Zulu 8.60.0.21-CA-linux64) (build 1.8.0_322-b06)
OpenJDK 64-Bit Server VM (Zulu 8.60.0.21-CA-linux64) (build 25.322-b06, mixed mode)
```

## Setting it up

```bash
# download it
cd /opt
sudo curl -o zulu8.60.0.21-ca-jdk8.0.322-linux_x64.tar.gz https://cdn.azul.com/zulu/bin/zulu8.60.0.21-ca-jdk8.0.322-linux_x64.tar.gz

# unzip it
sudo tar -xvf zulu8.60.0.21-ca-jdk8.0.322-linux_x64.tar.gz

# add it to your path file ~/.bashrc
PATH="$PATH:/opt/zulu8.60.0.21-ca-jdk8.0.322-linux_x64/bin"
```

## Example commands

### Starting the server

https://github.com/synle/rest.li/blob/master/restli-example-server/src/main/java/com/linkedin/restli/example/RestLiExampleBasicServer.java

```bash
./gradlew startExampleBasicServer
```

### Starting the client

https://github.com/synle/rest.li/blob/master/restli-example-client/src/main/java/com/linkedin/restli/example/RestLiExampleBasicClient.java

```bash
./gradlew startExampleBasicClient
```

### Basic java client code

For java client code, [refer to this sample client code in java for more information about java client code](https://github.com/synle/rest.li/blob/master/restli-example-client/src/main/java/com/linkedin/restli/example/RestLiExampleBasicClient.java). Essentially, we need to initiate

```java
import com.linkedin.restli.example.photos.PhotosRequestBuilders;
import com.linkedin.restli.client.RestClient;

//...
// this can be used to generate the request
private static final PhotosRequestBuilders _photoBuilders = new PhotosRequestBuilders();

// this used to make the actual rest client call
private final RestClient _restClient;

//...
// This section of the code initialize the rest client
public static void main(String[] args) throws Exception
{
  final StringBuilder serverUrlBuilder = new StringBuilder("http://").append(SERVER_HOSTNAME).append(":").append(SERVER_PORT).append("/");
  final RestClient restClient = new RestClient(r2Client, serverUrlBuilder.toString());
  //...
}

// here's a sample way to call the rest client
private void getNonPhoto() throws RemoteInvocationException
{
  final Request<Photo> failReq = _photoBuilders.get().id(-1L).build();
  final ResponseFuture<Photo> failFuture = _restClient.sendRequest(failReq);
  final Response<Photo> failResponse = failFuture.getResponse();
}
```

### Sample code and notes

- Resource file is located here: https://github.com/synle/rest.li/blob/master/restli-example-server/src/main/java/com/linkedin/restli/example/impl/PhotoResource.java
- Documentation regarding the resource can be found at https://linkedin.github.io/rest.li/user_guide/restli_server
- Note that the snapshot is generated automatically by code
  `rest.li/restli-example-api/src/main/snapshot/com.linkedin.restli.example.photos.photos.snapshot.json`
- All of these calls can be async. Just wrap the result into Task. So it's `Task<Map<K, V>>` instead of `Map<K, V>`

#### Get by ID

Annotate with this `@RestMethod.Get` or override the method as below:

##### Sample Server Java Code

```java
@Override
public Photo get(Long key)
{
  return _db.getData().get(key);
}
```

##### Sample Curl Call

```bash
curl http://localhost:7279/photos/1
```

##### Sample Server Response

```json
{
  "urn": "1",
  "format": "JPG",
  "id": 1,
  "title": "Photo 1",
  "exif": { "location": { "latitude": 66.7151, "longitude": -77.66235 } }
}
```

##### Sample Client Java Code

```java
// send request to retrieve created photo
private void getPhoto(PrintWriter respWriter, long newPhotoId) throws RemoteInvocationException
{
  final Request<Photo> getReq = _photoBuilders.get().id(newPhotoId).build();
  final ResponseFuture<Photo> getFuture = _restClient.sendRequest(getReq);
  final Response<Photo> getResp = getFuture.getResponse();
  respWriter.println("Photo: " + getResp.getEntity().toString());
}
```

#### Get All

Annotate with this `@RestMethod.BatchGet` or override the method as below:

##### Sample Server Java Code

```java
@Override
public List<Photo> getAll(@PagingContextParam PagingContext pagingContext)
{
  return new ArrayList<>(_db.getData().values());
}
```

##### Sample Curl Call

```bash
curl http://localhost:7279/photos
```

##### Sample Server Response

```json
{
  "elements": [
    {
      "urn": "1",
      "format": "PNG",
      "id": 1,
      "title": "Photo 1",
      "exif": { "location": { "latitude": 82.0381, "longitude": -64.79977 } }
    },
    {
      "urn": "2",
      "format": "$UNKNOWN",
      "id": 2,
      "title": "Photo 2",
      "exif": { "location": { "latitude": 61.78369, "longitude": 16.80307 } }
    },
    {
      "urn": "3",
      "format": "JPG",
      "id": 3,
      "title": "Photo 3",
      "exif": {
        "location": { "latitude": -30.974697, "longitude": -12.672791 }
      }
    },
    {
      "urn": "4",
      "format": "JPG",
      "id": 4,
      "title": "Photo 4",
      "exif": { "location": { "latitude": 0.5505829, "longitude": -60.278046 } }
    },
    {
      "urn": "5",
      "format": "JPG",
      "id": 5,
      "title": "Photo 5",
      "exif": { "location": { "latitude": 44.0925, "longitude": 71.784805 } }
    }
  ],
  "paging": {
    "count": 5,
    "start": 0,
    "links": [
      {
        "rel": "next",
        "type": "application/json",
        "href": "/photos?start=5&count=5"
      }
    ]
  }
}
```

##### Sample Client Java Code

```java
import com.linkedin.restli.client.GetAllRequest;

private void getAllPhotos(PrintWriter respWriter) throws RemoteInvocationException
{
  final GetAllRequest<Photo> getAllReq = _photoBuilders.getAll().build();
  final CollectionResponse<Photo> crPhotos = _restClient.sendRequest(getAllReq).getResponse().getEntity();
  final List<Photo> photos = crPhotos.getElements();

  respWriter.println("Get All Photo: " + photos.toString());
}
```

#### Batch Get

##### Sample Server Java Code

```java
@Override
public BatchResult<Long, Photo> batchGet(Set<Long> ids)
{
  System.out.println("\n\n>>>> batchGet:" + ids);

  Map<Long, Photo> result = new HashMap<>();
  Map<Long, RestLiServiceException> errors =
      new HashMap<>();

  for (Long key : ids)
  {
    if (get(key) != null)
    {
      result.put(key, get(key));
    }
    else
    {
      errors.put(key, new RestLiServiceException(HttpStatus.S_404_NOT_FOUND,
                                                 "No photo with id=" + key
                                                     + " has been found."));
    }
  }
  return new BatchResult<>(result, errors);
}
```

##### Sample Curl Call

```bash
curl http://localhost:7279/photos?ids=1&ids=2
```

##### Sample Server Response

```json
{"statuses":{},"results":{"1":{"urn":"1","format":"JPG","id":1,"title":"Photo 1","exif":{"location":{"latitude":46.466614,"longitude":-65.46433}}},"2":{"urn":"2","format":"$UNKNOWN","id":2,"title":"Photo 2","exif":{"location":{"latitude":66.07602,"longitude":64.918015}}}},"errors":{}}
```

##### Sample Client Java Code

```java
import com.linkedin.restli.client.BatchGetEntityRequest;
import com.linkedin.restli.common.EntityResponse;
import com.linkedin.restli.client.response.BatchKVResponse;
import java.util.Map;

private void batchGetAllPhotos(PrintWriter respWriter) throws RemoteInvocationException
{
  final BatchGetEntityRequest<Long, Photo> batchGetReq = _photoBuilders.batchGet().ids(1L).ids(2L).ids(3L).build();
  BatchKVResponse<Long,EntityResponse<Photo>> photoResponse = _restClient.sendRequest(batchGetReq).getResponse().getEntity();
  Map<Long, EntityResponse<Photo>> results = photoResponse.getResults();

  respWriter.println("Batch Get Photo: " + photoResponse.toString());

  EntityResponse<Photo> firstItem = results.get(1L);
  Photo firstPhoto = firstItem.getEntity();
  respWriter.println("Batch Get Photo - first photo: " + firstPhoto);
}
```

#### Update full

##### Sample Server Java Code

```java
@Override
public UpdateResponse update(Long key, Photo entity)
{
  System.out.println("\n\n>>>> update was called:" + key + ". " + entity);


  final Photo currPhoto = _db.getData().get(key);
  if (currPhoto == null)
  {
    return new UpdateResponse(HttpStatus.S_404_NOT_FOUND);
  }
  //Disallow changing entity ID and URN
  //ID and URN are required fields, so use a dummy value to denote "empty" fields
  if ((entity.hasId() && entity.getId() != -1) || (entity.hasUrn() && !entity.getUrn().equals("")))
  {
    throw new RestLiServiceException(HttpStatus.S_400_BAD_REQUEST,
                                     "Photo ID is not acceptable in request");
  }

  // make sure the ID in the entity is consistent with the key in the database
  entity.setId(key);
  entity.setUrn(String.valueOf(key));
  _db.getData().put(key, entity);
  return new UpdateResponse(HttpStatus.S_204_NO_CONTENT);
}
```

##### Sample Curl Call

```bash
curl -X PUT 'http://localhost:7279/photos/1' \
-H 'Content-Type: application/javascript' \
-d '{
    "title": "Photo 1a new name"
}'
```

##### Sample Server Response

```bash
HTTP/1.1 204 No Content
```

##### Sample Client Java Code

```java
private void createPhotoAsync(final PrintWriter respWriter, final CountDownLatch latch, final long newPhotoId)
{
  // this resembles to photo-create-id.json
  final LatLong newLatLong = new LatLong().setLatitude(40.725f).setLongitude(-74.005f);
  final EXIF newExif = new EXIF().setIsFlash(false).setLocation(newLatLong);
  final Photo newPhoto = new Photo().setTitle("Updated Photo").setFormat(PhotoFormats.JPG).setExif(newExif);

  final Request<EmptyRecord> createReq2 = _photoBuilders.update().id(newPhotoId).input(newPhoto).build();

  // send request with callback
  _restClient.sendRequest(createReq2);
}
```

#### Update Partial (Patch)

```java
@Override
public UpdateResponse update(Long key, PatchRequest<Photo> patchRequest)
{
  System.out.println("\n\n>>>> partial update (patch) was called:" + key + ". " + patchRequest);

  final Photo p = _db.getData().get(key);
  if (p == null)
  {
    return new UpdateResponse(HttpStatus.S_404_NOT_FOUND);
  }

  try
  {
    PatchApplier.applyPatch(p, patchRequest);
  }
  catch (DataProcessingException e)
  {
    return new UpdateResponse(HttpStatus.S_400_BAD_REQUEST);
  }
  // photo's id and URN should not be changed
  p.setId(key);
  p.setUrn(String.valueOf(key));
  _db.getData().put(key, p);

  return new UpdateResponse(HttpStatus.S_202_ACCEPTED);
}
```

##### Sample Curl Call

```bash
curl -X POST 'http://localhost:7279/photos/1' \
-H 'Content-Type: application/javascript' \
-d '{ "patch": { "$set": { "title": "Photo 1a with partial update name" } } }}'
```

##### Sample Server Response

```bash
HTTP/1.1 202 Accepted
```

##### Sample Client Java Code

```java
private void partialUpdatePhoto(PrintWriter respWriter, long photoId) throws RemoteInvocationException
{
  // get the original photo
  final Request<Photo> getReq = _photoBuilders.get().id(photoId).build();
  final ResponseFuture<Photo> getFuture = _restClient.sendRequest(getReq);
  final Response<Photo> getResp = getFuture.getResponse();
  final Photo originalPhoto = getResp.getEntity();

  // make a partial change and generate the diff
  final Photo updatedPhoto = new Photo().setTitle("Partially Updated Photo");
  final PatchRequest<Photo> patch = PatchGenerator.diff(originalPhoto, updatedPhoto);

  // send update request
  final Request<EmptyRecord> partialUpdateRequest = _photoBuilders.partialUpdate().id(photoId).input(patch).build();
  final int status = _restClient.sendRequest(partialUpdateRequest).getResponse().getStatus();
  respWriter.println("Partial update photo is successful: " + (status == 202));
}
```

#### Delete

##### Sample Server Java Code

```java
@Override

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

The process for developers to update the changelog is as follows:
- Always enter your change descriptions under the **"Unreleased"** heading.
  Do _not_ attempt to manually create your own version heading.
- When bumping the project version in your commit, run the `./scripts/update-changelog`
  script to automatically move everything under **"Unreleased"** to a new heading.

When updating the changelog, remember to be very clear about what behavior has changed
and what APIs have changed, if applicable.

## [Unreleased]

## [29.30.0] - 2022-02-12
- Support d2 config canary

## [29.29.1] - 2022-02-10
- Make DarkGateKeeper configurable for different dark clusters

## [29.29.0] - 2022-02-09
- Revert avro update introduced in 29.27.0

## [29.28.0] - 2022-02-04
- Fix weight double-count in D2 SimpleLoadBalancer

## [29.27.0] - 2022-01-25
- Update Avro version to 1.92

## [29.26.4] - 2022-01-24
- Map local variant of service ZNodes to cluster without colo suffix

## [29.26.3] - 2022-01-18
- Generate compile time constants for union members if people want to write build time switch case to check for things based on member

## [29.26.2] - 2022-01-13
- Fix for null pointer exception when registering listener on event bus before publisher is set.

## [29.26.1] - 2022-01-13
- Fail documentation requests while renderers are being lazily initialized (rather than block threads until complete).

## [29.26.0] - 2022-01-10
- Add header provider to generate required request to fetch default remote symbol table

## [29.25.0] - 2022-01-06
- Fix a race condition where TransportClient is shutdown while subset cache still holds reference to it

## [29.24.0] - 2021-12-09
- bump minor version for the new public method added in 29.23.3

## [29.23.3] - 2021-12-09
- Take default headers as input to fetch remote symbol table in `DefaultSymbolTableProvider`

## [29.23.2] - 2021-12-09
- Observability enhancements for D2 announcer.

## [29.23.1] - 2021-12-08
- Add support for a rate limiter supplier to enable multiple dark clusters with the CONSTANT_QPS strategy

## [29.23.0] - 2021-12-06
- Introduce d2ServiceName annotation on rest.li resources. This is an optional param that is meant to be populated by resources for whom the resource name is not the same as d2 service name

## [29.22.16] - 2021-12-03
- Fixed issues with potential duplicate TimingKeys being registered.

## [29.22.15] - 2021-11-30
- Add mock response generator factory for BATCH_FINDER methods.
- Deprecate `FileFormatDataSchemaParser#new(String, DataSchemaResolver, DataSchemaParserFactory)`. 
- Add file existence check before performing compatibility report check during snapshot and restmodel publishing

## [29.22.14] - 2021-11-24
- Fix bug where content type was not set for stream requests

## [29.22.13] - 2021-11-05
- Make SmoothRateLimiter setRate idempotent

## [29.22.12] - 2021-10-28
- add canaries to service and cluster properties

## [29.22.11] - 2021-10-25
- Fix an issue in D2 StateUpdater to force update PartitionState

## [29.22.10] - 2021-10-20
- SmoothRateLimiter - do not double-count execution delays on setRate

## [29.22.9] - 2021-10-12
- Make client timeout to fetch remote symbol table configurable in `RestLiSymbolTableProvider`

## [29.22.8] - 2021-10-12
- No changes; re-releasing because the previous release (`29.22.7`) was corrupted.

## [29.22.7] - 2021-10-11
- Fix bug for `generateDataTemplateTask` to consume command line option correctly.

## [29.22.6] - 2021-10-08
- Fix bug in `SmoothRateLimiter` where `executionDelay` is not honored.

## [29.22.5] - 2021-10-08
- Make `PegasusPlugin#getDataSchemaPath` public.

## [29.22.4] - 2021-09-28
- Improve support for JSR330 by allowing package protected constructors annotated with `@Inject`.
- Fix Supported mime type config for response payload.

## [29.22.3] - 2021-09-21
- Allow disabling of load balancing for a specific host.

## [29.22.2] - 2021-09-20
- Add server config support to define supported accept types

## [29.22.1] - 2021-09-13
- Mark the `extensions` directory as a resource root in the Gradle plugin.
- Add a pegasus plugin config to use case sensitive path in dataTemplate generation and rest client generation

## [29.22.0] - 2021-09-09
- Allow customizing `MethodAdapterRegistry` (now called `MethodAdapterProvider`) via `RestLiConfig`.
  - Rename `MethodAdapterRegistry` to `DefaultMethodAdapterProvider` and add interface `MethodAdapterProvider`.
  - Deprecate the constructors with `ErrorResponseBuilder` for `BaseRestLiServer` and its dependent classes
- Update the data template generator command-line app to accept a list of resolver directories
  to use for resolving schema references.
  - Also refactored the app to use a CLI library instead of passing arguments using system properties.
  - Update `GenerateDataTemplateTask` to use the refactored command-line app `DataTemplateGeneratorCmdLineApp`.
- `ConstantQpsDarkClusterStrategy` post-prod fixes.
  - Change the type of `dispatcherOutboundTargetRate` in `DarkClusterConfig.pdl` from `int` to `float`.
  - `ConstantQpsRateLimiter` - Introduce randomness while maintaining constant per-period rate.

## [29.21.5] - 2021-09-09
- Fix a bug in `DataTranslator` where accessing non-existent fields under avro 1.10+ throws an exception.

## [29.21.4] - 2021-08-30
- Expose an API to build a URI without query params. Expose a local attr for passing query params for in-process calls.

## [29.21.3] - 2021-08-25
- Fix a bug in `SmoothRateLimiter` where `getEvents` will always return `0`.

## [29.21.2] - 2021-08-18
- Remove support for disabling request validation via headers since doing so can have dangerous side effects.

## [29.21.1] - 2021-08-18

restli-example-client/src/main/java/com/linkedin/restli/example/RestLiExampleBasicClient.java
/*
   Copyright (c) 2012 LinkedIn Corp.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

package com.linkedin.restli.example;


import com.linkedin.common.callback.Callback;
import com.linkedin.common.callback.FutureCallback;
import com.linkedin.common.util.None;
import com.linkedin.r2.RemoteInvocationException;
import com.linkedin.r2.transport.common.Client;
import com.linkedin.r2.transport.common.bridge.client.TransportClient;
import com.linkedin.r2.transport.common.bridge.client.TransportClientAdapter;
import com.linkedin.r2.transport.http.client.HttpClientFactory;
import com.linkedin.restli.client.FindRequest;
import com.linkedin.restli.client.GetAllRequest;
import com.linkedin.restli.client.DeleteRequest;
import com.linkedin.restli.client.BatchGetEntityRequest;
import com.linkedin.restli.client.Request;
import com.linkedin.restli.client.Response;
import com.linkedin.restli.client.ResponseFuture;
import com.linkedin.restli.client.RestClient;
import com.linkedin.restli.client.util.PatchGenerator;
import com.linkedin.restli.common.CollectionResponse;
import com.linkedin.restli.common.EmptyRecord;
import com.linkedin.restli.common.IdResponse;
import com.linkedin.restli.common.PatchRequest;
import com.linkedin.restli.example.photos.AlbumEntryRequestBuilders;
import com.linkedin.restli.example.photos.AlbumsRequestBuilders;
import com.linkedin.restli.example.photos.PhotosRequestBuilders;
import com.linkedin.restli.common.EntityResponse;
import com.linkedin.restli.client.response.BatchKVResponse;

import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Random;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.Map;


/**
 * @author Keren Jin
 */
public class RestLiExampleBasicClient
{
  /**
   * This is a stand-alone app to demo the use of client-side Pegasus API. To run in,
   * com.linkedin.restli.example.RestLiExamplesServer has to be running.
   *
   * The only argument is the path to the resource on the photo server, e.g. /photos/1
   */
  public static void main(String[] args) throws Exception
  {
    // create HTTP Netty client with default properties
    final HttpClientFactory http = new HttpClientFactory.Builder().build();
    final TransportClient transportClient = http.getClient(Collections.<String, String>emptyMap());
    // create an abstraction layer over the actual client, which supports both REST and RPC
    final Client r2Client = new TransportClientAdapter(transportClient);
    // REST client wrapper that simplifies the interface
    final StringBuilder serverUrlBuilder = new StringBuilder("http://").append(SERVER_HOSTNAME).append(":").append(SERVER_PORT).append("/");
    final RestClient restClient = new RestClient(r2Client, serverUrlBuilder.toString());
    final RestLiExampleBasicClient photoClient = new RestLiExampleBasicClient(restClient);

    String pathInfo = args.length == 0 ? "" : args[0];
    photoClient.sendRequest(pathInfo, new PrintWriter(System.out));
    photoClient.shutdown();
    http.shutdown(new FutureCallback<>());
  }

  public RestLiExampleBasicClient(RestClient restClient)
  {
    _restClient = restClient;
  }

  public void sendRequest(String pathInfo, PrintWriter respWriter)
  {
    /* Supported URLs
     *  /fail: just fail
     *  /album/<album_id>: display album
     *  all others: make photos, get photos, purge photos
     */
    try
    {
      if (pathInfo.equals("/fail"))
      {
        getNonPhoto();
      }
      else
      {
        if (pathInfo.startsWith("/album/"))
        {
          getAlbum(respWriter, Long.parseLong(pathInfo.substring("/album/".length())));
        }
        else
        {
          // this track does not make any assumption on server
          // we need to create photo first, find it and clean them up
          // we use both sync and async approaches for creating

          final long newPhotoId = createPhoto(respWriter);
          final CountDownLatch latch = new CountDownLatch(1);
          createPhotoAsync(respWriter, latch, newPhotoId);
          getPhoto(respWriter, newPhotoId);
          getAllPhotos(respWriter);
          batchGetAllPhotos(respWriter);
          deletePhoto(respWriter);
          findPhoto(respWriter);
          partialUpdatePhoto(respWriter, newPhotoId);
          // photos and albums have IDs starting from 1
          getAlbumSummary(respWriter, (long) new Random().nextInt(10) + 1);
          purgeAllPhotos(respWriter);
          try
          {
            latch.await();
          }
          catch (InterruptedException e)
          {
            respWriter.println(e.getMessage());
          }
        }
      }
    }
    catch (RemoteInvocationException e)
    {
      respWriter.println("Error in example client: " + e.getMessage());
    }
    respWriter.flush();
  }

  public void shutdown()
  {
    _restClient.shutdown(new Callback<None>()
    {
      @Override
      public void onError(Throwable e)
      {
        throw new RuntimeException("Error occurred during example client shutdown.", e);
      }

      @Override
      public void onSuccess(None result)
      {
        System.out.println("Example client is shutdown.");
      }
    });
  }

  private long createPhoto(PrintWriter respWriter) throws RemoteInvocationException
  {
    // make create photo request and send with the rest client synchronously
    // response of create request does not have body, therefore use EmptyRecord as template

    // create an instance of photo pragmatically
    // this resembles to photo-create.json
    final LatLong newLatLong = new LatLong().setLatitude(37.42394f).setLongitude(-122.0708f);
    final EXIF newExif = new EXIF().setLocation(newLatLong);
    final Photo newPhoto = new Photo().setTitle("New Photo").setFormat(PhotoFormats.PNG).setExif(newExif);

    final Request<IdResponse<Long>> createReq1 = _photoBuilders.create().input(newPhoto).build();
    final ResponseFuture<IdResponse<Long>> createFuture1 = _restClient.sendRequest(createReq1);
    // Future.getResource() blocks until server responds
    final Response<IdResponse<Long>> createResp1 = createFuture1.getResponse();
    createResp1.getEntity();

    @SuppressWarnings("unchecked")
    final IdResponse<Long> entity = createResp1.getEntity();
    final long newPhotoId = entity.getId();
    respWriter.println("New photo ID: " + newPhotoId);

    return newPhotoId;
  }

  /**
   * make create photo request and send with the rest client asynchronously
   */
  private void createPhotoAsync(final PrintWriter respWriter, final CountDownLatch latch, final long newPhotoId)
  {
    // this resembles to photo-create-id.json
    final LatLong newLatLong = new LatLong().setLatitude(40.725f).setLongitude(-74.005f);
    final EXIF newExif = new EXIF().setIsFlash(false).setLocation(newLatLong);
    final Photo newPhoto = new Photo().setTitle("Updated Photo").setFormat(PhotoFormats.JPG).setExif(newExif);

    final Request<EmptyRecord> createReq2 = _photoBuilders.update().id(newPhotoId).input(newPhoto).build();

    // send request with callback
    _restClient.sendRequest(createReq2, new Callback<Response<EmptyRecord>>()
    {
      @Override
      public void onError(Throwable e)
      {
        respWriter.println(e.getMessage());
        latch.countDown();
      }

      @Override
      public void onSuccess(Response<EmptyRecord> result)
      {
        respWriter.println("Update photo is successful: " + (result.getStatus() == 204));

        // without a condition variable or CountDownLatch,
        // this callback might not be called before main thread terminates
        latch.countDown();
      }
    });
  }

  /**
   * Retrieve the album information and each photo in the album. The photos are retrieved in parallel.
   */
  private void getAlbum(PrintWriter respWriter, long albumId) throws RemoteInvocationException
  {
    // get the specific album
    final Request<Album> getAlbumReq = _albumBuilders.get().id(albumId).build();
    final ResponseFuture<Album> getAlbumFuture = _restClient.sendRequest(getAlbumReq);
    final Response<Album> getResp = getAlbumFuture.getResponse();
    final Album album = getResp.getEntity();

    respWriter.println(album.getTitle());
    respWriter.println("Created on " + new Date(album.getCreationTime()));

    // get the album's entries
    final FindRequest<AlbumEntry> searchReq = _albumEntryBuilders.findBySearch().albumIdParam(albumId).build();
    final ResponseFuture<CollectionResponse<AlbumEntry>> responseFuture = _restClient.sendRequest(searchReq);
    final Response<CollectionResponse<AlbumEntry>> response = responseFuture.getResponse();
    final List<AlbumEntry> entries = new ArrayList<>(response.getEntity().getElements());

    entries.add(new AlbumEntry().setAlbumId(-1).setPhotoId(9999));

    // don't return until all photo requests done
    final CountDownLatch latch = new CountDownLatch(entries.size());

    // fetch every photo asynchronously
    // store either a photo or an exception
    final Object[] photos = new Object[entries.size()];
    for (int i = 0; i < entries.size(); i++)
    {
      final int finalI = i; // need final version for callback
      final AlbumEntry entry = entries.get(i);
      final long photoId = entry.getPhotoId();
      final Request<Photo> getPhotoReq = _photoBuilders.get().id(photoId).build();
      _restClient.sendRequest(getPhotoReq, new Callback<Response<Photo>>()
      {
        @Override
        public void onSuccess(Response<Photo> result)
        {
          photos[finalI] = result.getEntity();
          latch.countDown();
        }

        @Override
        public void onError(Throwable e)
        {
          photos[finalI] = e;
        }
      });
    }

    try
    {
      // wait for all requests to finish
      latch.await(2, TimeUnit.SECONDS);
      if (latch.getCount() > 0)
      {
        respWriter.println("Failed to retrieve some photo(s)");
      }
    }
    catch (InterruptedException e)
    {
      e.printStackTrace(respWriter);
    }

    // print photo data
    for (int i = 0; i < entries.size(); i++)
    {
      final Object val = photos[i];
      final AlbumEntry entry = entries.get(i);
      if (val instanceof Throwable)
      {
        respWriter.println("Failed to load photo " + entry.getPhotoId());
        respWriter.println("Stack trace:");
        ((Throwable) val).printStackTrace(respWriter);
        respWriter.println();
      }
      else if (val instanceof Photo)
      {
        final Photo photo = (Photo) val;
        respWriter.println("Photo " + photo.getTitle() + ":");
        respWriter.println(photo);
        respWriter.println("Added on " + new Date(entry.getAddTime()));
      }
      else
      {
        throw new AssertionError("expected photo or exception");
      }
    }
  }

  /**
   * send request to retrieve created photo
   */
  private void getPhoto(PrintWriter respWriter, long newPhotoId) throws RemoteInvocationException
  {
    final Request<Photo> getReq = _photoBuilders.get().id(newPhotoId).build();
    final ResponseFuture<Photo> getFuture = _restClient.sendRequest(getReq);
    final Response<Photo> getResp = getFuture.getResponse();
    respWriter.println("Photo: " + getResp.getEntity().toString());
  }

  private void getAllPhotos(PrintWriter respWriter) throws RemoteInvocationException
  {
    final GetAllRequest<Photo> getAllReq = _photoBuilders.getAll().build();
    final CollectionResponse<Photo> crPhotos = _restClient.sendRequest(getAllReq).getResponse().getEntity();
    final List<Photo> photos = crPhotos.getElements();

    respWriter.println("Get All Photo: " + photos.toString());
  }

  private void batchGetAllPhotos(PrintWriter respWriter) throws RemoteInvocationException
  {
    final BatchGetEntityRequest<Long, Photo> batchGetReq = _photoBuilders.batchGet().ids(1L).ids(2L).ids(3L).build();
    BatchKVResponse<Long,EntityResponse<Photo>> photoResponse = _restClient.sendRequest(batchGetReq).getResponse().getEntity();
    Map<Long, EntityResponse<Photo>> results = photoResponse.getResults();

    respWriter.println("Batch Get Photo: " + photoResponse.toString());

    EntityResponse<Photo> firstItem = results.get(1L);
    Photo firstPhoto = firstItem.getEntity();
    respWriter.println("Batch Get Photo - first photo: " + firstPhoto);
  }

  private void deletePhoto(PrintWriter respWriter) throws RemoteInvocationException {
    final DeleteRequest<Photo> deleteRequest = _photoBuilders.delete().id(5L).build();
    final int status = _restClient.sendRequest(deleteRequest).getResponse().getStatus();
    respWriter.println("Delete Photo: StatusCode=" + status);
  }


  /**
   * call action purge to delete all photos on server
   */
  private void purgeAllPhotos(PrintWriter respWriter) throws RemoteInvocationException
  {
    final Request<Integer> purgeReq = _photoBuilders.actionPurge().build();
    final ResponseFuture<Integer> purgeFuture = _restClient.sendRequest(purgeReq);
    final Response<Integer> purgeResp = purgeFuture.getResponse();
    respWriter.println("Purged " + purgeResp.getEntity() + " photos");
  }



  /**
   * retrieve album
   */
  private void getAlbumSummary(PrintWriter respWriter, long newAlbumId) throws RemoteInvocationException
  {
    final Request<Album> getReq = _albumBuilders.get().id(newAlbumId).build();
    final ResponseFuture<Album> getFuture = _restClient.sendRequest(getReq);
    final Response<Album> getResp = getFuture.getResponse();
    respWriter.println("Album: " + getResp.getEntity().toString());
  }

  /**
   * failed request that try to access non-existing photo and throw RestException
   */
  private void getNonPhoto() throws RemoteInvocationException
  {
    final Request<Photo> failReq = _photoBuilders.get().id(-1L).build();
    final ResponseFuture<Photo> failFuture = _restClient.sendRequest(failReq);
    final Response<Photo> failResponse = failFuture.getResponse();
  }

  /**
   *
   */
  private void partialUpdatePhoto(PrintWriter respWriter, long photoId) throws RemoteInvocationException
  {
    final Request<Photo> getReq = _photoBuilders.get().id(photoId).build();
    final ResponseFuture<Photo> getFuture = _restClient.sendRequest(getReq);
    final Response<Photo> getResp = getFuture.getResponse();
    final Photo originalPhoto = getResp.getEntity();

    final Photo updatedPhoto = new Photo().setTitle("Partially Updated Photo");
    final PatchRequest<Photo> patch = PatchGenerator.diff(originalPhoto, updatedPhoto);

    final Request<EmptyRecord> partialUpdateRequest = _photoBuilders.partialUpdate().id(photoId).input(patch).build();
    final int status = _restClient.sendRequest(partialUpdateRequest).getResponse().getStatus();
    respWriter.println("Partial update photo is successful: " + (status == 202));
  }

  /**
   * use Finder to find the photo with some criteria
   */
  private void findPhoto(PrintWriter respWriter) throws RemoteInvocationException
  {
    final long newPhotoId = createPhoto(respWriter);
    createPhoto(respWriter);
    createPhoto(respWriter);

    final Request<Photo> getReq = _photoBuilders.get().id(newPhotoId).build();
    final ResponseFuture<Photo> getFuture = _restClient.sendRequest(getReq);
    final Response<Photo> getResp = getFuture.getResponse();
    final Photo photo = getResp.getEntity();

    final FindRequest<Photo> findReq = _photoBuilders
        .findByTitleAndOrFormat()
        .titleParam(photo.getTitle())
        .formatParam(photo.getFormat())
        .build();

    final CollectionResponse<Photo> crPhotos = _restClient.sendRequest(findReq).getResponse().getEntity();
    final List<Photo> photos = crPhotos.getElements();

    respWriter.println("Found " + photos.size() + " photos with title " + photo.getTitle());
  }

  private static final String SERVER_HOSTNAME = "localhost";
  private static final int SERVER_PORT = 7279;

  // builder is convenient way to generate rest.li request
  private static final PhotosRequestBuilders _photoBuilders          = new PhotosRequestBuilders();
  private static final AlbumsRequestBuilders _albumBuilders          = new AlbumsRequestBuilders();
  private static final AlbumEntryRequestBuilders _albumEntryBuilders = new AlbumEntryRequestBuilders();

  private final RestClient _restClient;
}
